### ISSUE TITLE
Assertions with user-defined error codes

### ISSUE THREAD

**Issue by bobbinth** [0xPolygonMiden/miden-vm#988]
Currently, when program execution fails due to a violated assertion, the VM reports just that an assertion has failed at a specific cycle. This makes it rather difficult to find out what exactly went wrong. One way to help with diagnostics is to return an error code together with the `FailedAssertion` error.

Overall, we have 7 assertion instructions which we could enable this functionality for: `assert`, `assertz`, `assert_eq`, `assert_eqw`, `u32assert.1`, `u32assert.2`, `u32assertw`.

On the assembly side, the syntax could be similar to our usual syntax for parameters. For example:
```
assert      # regular assert instruction with no error code
assert.123  # assert instruction with error code 123
```
One downside of the above format is that it may be misinterpreted that we are asserting that top of the stack is `123`. At the expense of a slightly more complex parsing, we could make things more explicit:
```
assert.err=123
```

Both approaches will work fine for all relevant instructions except for `u32assert.1` and `u32assert.2`. One way to deal with this is to rename them as follows:
- `u32assert.1` -> `u32assert`.
- `u32assert.2` -> `u32assert2`.

I think this renaming is a good idea in and of itself.

**Comment by hackaugusto** [comment]
It would be best to support error messages with formatting. When the assert in inside a loop the error code is helpful, but still not sufficient to determine which values caused a problem.

Having an error message is also useful for users besides the original author (the user could be another developer, or even the end-user). Given that the error message is written somewhat nicely.

We could have formatting done in miden assembly itself, something like:

`u32assert."Trace length can not be larger than a u32 value"

If we add formatting, we could have something like:

`u32assert."Trace length can not be larger than a u32 value, got {0}"

Where `{0}` is stack offset to be formatted in the error message. Although that would have value for error description, it would have runtime cost. I guess we could also allow referring to old stack values, since we have the data in the trace anyways, at the cost of a surprisingly complicated format string.

**Comment by bobbinth** [comment]
Agreed. Though, I'd rather start with something simple and then expand from there. For example, for strings, we'd need to update our parsing logic as we treat whitespace as instruction separators.

**Comment by frisitano** [comment]
What's the priority on this issue? This is of particular relevance when working with and debugging transaction programs in `miden-base`.  I think it would be beneficial to get this in as soon as possible.

**Comment by bobbinth** [comment]
No concrete progress yet - but if we prioritize it, I think we can get it done next week.

**Comment by frisitano** [comment]
I think we should prioritise this. It's not only beneficial for reasons described above but also for testing failure cases when we want to assert some code (tx kernel) fails an assertion for a particular reason.

**Comment by frisitano** [comment]
An additional suggestion would be to support constants e.g. assert.err=ASSET_INVARIANT_ERROR

**Comment by bobbinth** [comment]
> I think we should prioritise this. 

Agreed. There is one other thing we could do, and I'm wondering if it would be more useful for debugging purposes than assert with custom error codes (we could do both, just a question of priorities): add a `debug` decorator. We actually used to have something like this but removed it because it complicated things in the old approach to decorators. The approach has been refactored since then - so, having it shouldn't complicate things any more.

The decorator used to work as follows:

- `debug.all` prints out the entire state of the VM (stack and RAM).
- `debug.stack` prints out the entire contents of the stack.
- `debug.stack.<n>` prints out the top $n$ items of the stack. $n$ must be an integer greater than $0$.
- `debug.mem` prints out the entire contents of RAM.
- `debug.mem.<n>` prints out contents of memory at address $n$.
- `debug.mem.<n>.<m>` prints out the contents of memory starting at address $n$ and ending at address $m$ (both inclusive). $m$ must be greater than $n$.
- `debug.local` prints out all locals of the currently executing procedure.
- `debug.local.<n>` prints out contents of the local at index $n$ for the currently executing procedure.

**Comment by frisitano** [comment]
I agree that the `debug` decorator would certainly be useful.  However, I think the user-defined error assertion should be prioritised for the following reason.  Say we are writing some code which leverages some library.  We are not the owners of this library and as such do not know much about its internals.  Upon execution of the code an assertion fails and this failure arises from within the library code.  We do not know why this assertion is failing and we also do not know where in the library code the failed assertion is coming from.  Having user-defined error codes would allow the user to reference the error code against the error spec and quickly identify why and where in the library source code the assertion is failing.  Once we have this location it would then be beneficial to inject a `debug` decorator to extract information relating to the state of the vm at or just before the assertion fails. 

This is of particular relevance for clients using the transaction kernel to construct transactions using user defines notes and transaction scripts.  In some instances assertions within the kernel will fail and the client will have to dive into the transaction kernel code to try and understand why this is happening.

**Comment by bobbinth** [comment]
Yes, good points! But I've actually already implemented a basic `debug` instruction in #1069 (in part because I think it is a bit more straight-forward). I'll tackle `assert` next, and hopefully, it'll be done in the next couple of days.

**Comment by frisitano** [comment]
> Yes, good points! But I've actually already implemented a basic `debug` instruction in #1069 (in part because I think it is a bit more straight-forward). I'll tackle `assert` next, and hopefully, it'll be done in the next couple of days.

Awesome, thanks for this! Lets start thinking about a standard for the definition of error codes. Given then we don't have first class support for constant exports I think we would need to define them at the top of the source file they are raised in.  This is obviously dependent on having constant support for the assert with user defined error but I think this is quite important.

**Comment by bobbinth** [comment]
Most of this has been implemented in #1075. The two things that remain are: `u32assert.1` and `u32assert.2`. For these operations, we should first rename them as follows:

- `u32assert.1` -> `u32assert`
- `u32assert.2` -> `u32assert2`

And then add implement custom error messages for these in the same was as we did for other instructions.

Another thing we should probably define is error code format. The main purpose of this would be to reduce potential for the same error code being defined in multiple libraries. One potential way to do it is as follows:

- Use the first (or the last) 2 bytes of the error code to define a library. For example, errors coming from assembly instruction would have the value of the first two bytes set to 0. Errors coming from `stdlib` would have this value set to `1`. Other libraries would be free to pick error code prefixes. In the future, we may want to maintain a list of well-known libraries together with error code prefixes they use.
- The last (or the first) 2 bytes would be used for a specific error code.

The above scheme gives us roughly 65K unique values for library prefixes and 65K unique error codes per library.
